### ISSUE TITLE
Best practice to pass dynamic parameters to an UseCase?

### ISSUE THREAD

**Issue by rockerhieu** [android10/Android-CleanArchitecture#32]
As in the example, usecase's parameters are set via `UserModule`:

``` java
@Module
public class UserModule {
  private int userId = -1;
  public UserModule() {}

  public UserModule(int userId) {
    this.userId = userId;
  }

  @Provides @PerActivity @Named("userDetails") UseCase provideGetUserDetailsUseCase(
      UserRepository userRepository, ThreadExecutor threadExecutor,
      PostExecutionThread postExecutionThread) {
    return new GetUserDetailsUseCase(userId, userRepository, threadExecutor, postExecutionThread);
  }
}
```

But there are some cases when `buildUseCaseObservable` depends on some dynamic parameters. I tried to create an interface to provide these parameters and let the view (in MVP) implement it and pass them to `UserModule`. With this approach if the view is a `Fragment` then I have to re-create `UserModule` again in the `Fragment`.

``` java
public interface UserIdProvider {
  int getUserId();
}
```

Any suggestion, recommendation?

**Comment by android10** [comment]
So in that case, buildUserCaseObservable is no longer useful and you will have to pass them via setter (with defensive code and failing fast in case all the dynamic parameters are not satisfied). 
I used this approach due to dagger instantiating my objects thus, being able to pass the user id via module constructor parameter.

**Comment by spirosoik** [comment]
@android10 yes that's true, but how do you pass dynamics parameters for login process for example? I mean in this case there isn't a predefined parameter but you must setup the submitted details? Which is your approach on this? I know that this is a question but it's really interesting to give us an example.

For this case the Dagger module must be dynamic so I suppose that there is two solutions (I think):
1) PerFragment modules
2) Observer to get the changes and re-initialises  the Activity dagger module.

or ????

**Comment by spirosoik** [comment]
@android10 please when you have free time it will be nice to give us a better approach by your side.

**Comment by spirosoik** [comment]
@android10 I know that is closed this issues but I want to know if our approach is acceptable.

Currently we have multiple forms so for example I have an invite form with multiple attrs.

Let's Say I have  ```InviteActivity``` ```InviteView``` ```InvitePresenter``` ```InviteComponent``` in their referral package explicitly.

In my ```InviteModule``` of dagger I have something like this:

```java
/**
 * Dagger module that provides user related collaborators.
 */
@Module public class InviteModule {

  private String email;
  private String businessName;

  public InviteModule() {
  }

  public InviteModule(String email, String businessName) {
    this.email = email;
    this.businessName = businessName;
  }

  @Provides @PerActivity @Named("userLogin") UseCase provideInviteBusinessUseCase(InviteBusiness sendInviteBusiness) {
    return new InviteBusiness(this.email, this.businessName, UserRepository, threadExecutor,
        postExecutionThread);
  }
}
```

**Comment by MehdiChouag** [comment]
@spirosoik Does the solution example in your last comment works nice with your needs ?

**Comment by spirosoik** [comment]
@MehdiChouag It's easier to keep the abstraction as it is with ```buildUseCaseObservable()`` but you can use this. Remember that in this method you must pass a Domain model not simple parameters.

```  protected abstract Observable buildUseCaseObservable(Object...param);```

**Comment by MehdiChouag** [comment]
@spirosoik For me the first solution seems to be more appropriate.

Rather using `Object`, we may using template for parameters, like : 

```java
public abstract class UseCase<T> {
...
protected abstract Observable buildUseCaseObservable(@Nullable T... params);

public void execute(Subscriber UseCaseSubscriber, @Nullable T... params) {
    mSubscription = buildUseCaseObservable(params).subscribeOn(mExecutionThread.getScheduler())
        .observeOn(mPostExecutionThread.getScheduler())
        .subscribe(UseCaseSubscriber);
  }
}
```
And using it like this :
```java
public class LoginAccountUseCase extends UseCase<String> {
@Override
  protected Observable buildUseCaseObservable(String... params) {
    return mAccountRepository.login(params[0], params[1]);
  }
}
```

**Comment by spirosoik** [comment]
@MehdiChouag Yes of course I just send you the logic. This is depends in your case but as I said before it will be nice to use domain models to use cases, not plain params.

For example I am passing into the UseCase a UserModel which I am transforming this into User domain model or for credentials you can have a different object for this.

**Comment by MehdiChouag** [comment]
@spirosoik Can we avoid to create a UserModel then transforming into a User domain ?
And instead directly use the User domain, that wouldn't break dependency rules.

**Comment by spirosoik** [comment]
@MehdiChouag Yes there is a great discussion (you are in) on this subject and I will support @android10 on this that each layer must have the View Models and each layer should have its model to deal with in order to avoid coupling between layers.

**Comment by jatago** [comment]
Let me give my 2 cents here. 

@MehdiChouag in reality you are not breaking any dependency rule. The domain layer does not use any model from the presentation layer but the opposite. This would leave the dependency as follows:

**Comment by Fshamri** [comment]
@jatago, so init must be abstract in UseCase?

**Comment by jatago** [comment]
In my case I don't actually include it in the base UseCase as abstract method but do it case by case. There might be some cases in which I required different params and some cases in which I don't require any. I just make sure to throw an early exception if any of the required params for this particular UseCase are not set and good to go.

**Comment by astelmakh** [comment]
@jatago what to do with a hard dependency UserLogin in Presentation layer in this case ?

**Comment by lutaseb** [comment]
Sorry to interrupt here but this a matter of architecture, I referenced many issues
- execute() can take more parameters than just a subscriber witch means buildUseCaseObservable should be overrided as well to pass the defined parameters ...so the base class is almost useless
btw, injecting UseCase instances when the execute() method doesn't have the same signature is kind of ugly I guess
- the subscribtion has to be unsubscribe before calling another execute() with other parameters (if the useCase is the same instance of course)
I think a useCase should should only be reused if a pagination is made, nothing more
- creating a dagger module seems the "cleanest" way to me but seriously this is overkill
....still I don't have any answer but the issue is really serious

**Comment by sergio11** [comment]
It would be a good solution pass bundle to buildUseCaseObservable method?

```
public class SigninInteractor extends BaseInteractor {

    public final static String EMAIL_PARAM = "email_param";
    public final static String PASSWORD_PARAM = "password_param";

    private FirebaseAuth firebaseAuth;

    public SigninInteractor(FirebaseAuth firebaseAuth) {
        this.firebaseAuth = firebaseAuth;
    }

    @Override
    protected Observable<AuthResult> buildUseCaseObservable(Bundle bundle) {
        if(bundle == null){
            throw new Error("Invalid Params");
        }
        return RxFirebaseAuth.signInWithEmailAndPassword(firebaseAuth, bundle.getString(EMAIL_PARAM), bundle.getString(PASSWORD_PARAM));
    }
}
```

**Comment by roscrazy** [comment]
Hi all,

I know this thread have closed but I want to share my approach, please leave some comment if possible :
In my approach, I create a factory for the UseCase, and when needed, I will create a new UseCase in Presenter from the Factory.
For example I have this UseCase :
`public class DeleteFeedUseCase extends UseCase {

    @NonNull
    FeedRepository mRepository;

    @NonNull
    String mFeedKey;

    @Inject
    public DeleteFeedUseCase(String  feed, FeedRepository repository, ThreadExecutor threadExecutor, PostExecutionThread postExecutionThread) {
        super(threadExecutor, postExecutionThread);
        this.mRepository = repository;
        this.mFeedKey = feed;
    }

    @Override
    protected Observable buildUseCaseObservable() {
        return mRepository.deleteFeed(mFeedKey);
    }
}`

And this is the Factory :

`public class DeleteFeedUseCaseFactory {
    private FeedRepository mRepository;
    private ThreadExecutor mThreadExecutor;
    private PostExecutionThread mPostExecutionThread;

    @Inject
    public DeleteFeedUseCaseFactory(FeedRepository mRepository, ThreadExecutor mThreadExecutor, PostExecutionThread mPostExecutionThread) {
        this.mRepository = mRepository;
        this.mThreadExecutor = mThreadExecutor;
        this.mPostExecutionThread = mPostExecutionThread;
    }

    public DeleteFeedUseCase create(String feed){
        return new DeleteFeedUseCase(feed, mRepository, mThreadExecutor, mPostExecutionThread);
    }
}`

And how it is used in presenter :

   `    void deleteFeed(int index){
        String key = mKeyStore.get(index);

        DeleteFeedUseCase deleteFeedUseCase = mDeleteFeedUseCaseFactory.create(key);

        deleteFeedUseCase.execute(new DeleteFeedSubscriber());
        unsubscribeOnUnbindView(deleteFeedUseCase);
    }`

You can check out more detail on my sample https://github.com/roscrazy/Android-RealtimeUpdate-CleanArchitecture

**Comment by astelmakh** [comment]
`public abstract class UseCase<P extends UseCase.RequestValues, Q extends UseCase.ResponseValues> {`private P requestValues; 
private P requestValues;


    private P requestValues;protected ThreadExecutor threadExecutor;
    protected PostExecutionThread postExecutionThread;

    private Subscription subscription = Subscriptions.empty();

    public UseCase(ThreadExecutor threadExecutor, PostExecutionThread postExecutionThread) {
        this.threadExecutor = threadExecutor;
        this.postExecutionThread = postExecutionThread;
    }

    protected P getRequestValues() {
        return requestValues;
    }

    protected abstract Observable<Q> buildObservable(P requestValues);

    public final void executeUseCase(DefaultSubscriber<Q> subscriber, P requestValues) {
        this.subscription = checkNotNull(buildObservable(requestValues), "Observable cant be null")
                .compose(applySchedulers())
                .subscribe(subscriber);
    }

    public void unsubscribe() {
        if(!subscription.isUnsubscribed())
           subscription.unsubscribe();
    }

    protected final <T> Observable.Transformer<T, T> applySchedulers() {
        return observable -> observable.subscribeOn(Schedulers.from(threadExecutor))
                .observeOn(postExecutionThread.getScheduler());
    }

    public interface RequestValues {
    }

    public interface ResponseValues {
    }
}`

**Comment by astelmakh** [comment]
Try this.

```
public abstract class UseCase<P extends UseCase.RequestValues, Q extends UseCase.ResponseValues> {
    private P requestValues;

    protected ThreadExecutor threadExecutor;
    protected PostExecutionThread postExecutionThread;

    private Subscription subscription = Subscriptions.empty();

    public UseCase(ThreadExecutor threadExecutor, PostExecutionThread postExecutionThread) {
        this.threadExecutor = threadExecutor;
        this.postExecutionThread = postExecutionThread;
    }

    protected P getRequestValues() {
        return requestValues;
    }

    protected abstract Observable<Q> buildObservable(P requestValues);

    public final void executeUseCase(DefaultSubscriber<Q> subscriber, P requestValues) {
        this.subscription = checkNotNull(buildObservable(requestValues), "Observable cant be null")
                .compose(applySchedulers())
                .subscribe(subscriber);
    }

    public void unsubscribe() {
        if(!subscription.isUnsubscribed())
           subscription.unsubscribe();
    }

    protected final <T> Observable.Transformer<T, T> applySchedulers() {
        return observable -> observable.subscribeOn(Schedulers.from(threadExecutor))
                .observeOn(postExecutionThread.getScheduler());
    }

    public interface RequestValues {
    }

    public interface ResponseValues {
    }
}
```

```
public class GetProfileUseCase extends UseCase<GetProfileUseCase.RequestValues,
        GetProfileUseCase.ResponseValues> {

    private ProfileRepository profileRepository;

    public GetProfileUseCase(ThreadExecutor threadExecutor,
                             PostExecutionThread postExecutionThread,
                             ProfileRepository profileRepository) {
        super(threadExecutor, postExecutionThread);
        this.profileRepository = profileRepository;
    }

    @Override
    protected Observable<ResponseValues> buildObservable(RequestValues requestValues) {
        return profileRepository.get(requestValues.getProfileId())
                .map(profile -> new ResponseValues(profile));
    }

    public static final class RequestValues implements UseCase.RequestValues {
        private final long profileId;

        public RequestValues(long profileId) {
            this.profileId = profileId;
        }

        public long getProfileId() {
            return profileId;
        }
    }

    public static final class ResponseValues implements UseCase.ResponseValues {

        private final Profile profile;

        public ResponseValues(Profile profile) {
            this.profile = profile;
        }

        public Profile getProfile() {
            return profile;
        }
    }
}

```

**Comment by roscrazy** [comment]
Hi @astelmakh,
I checked your solution. Could it be any problem with the usecase's subscription if we run the "executeUseCase()" multiple time? Example call "executeUseCase" 5 times immediately.

**Comment by android10** [comment]
I re-opened this since it was a pending task and due to the lack of time I could not do it. 
Thanks for all the input/feedback, here is the commit that allows us to pass dynamic parameters to our use cases: https://github.com/android10/Android-CleanArchitecture/commit/dfd61c29170b74298780ebb7cc991e0b2edbdfe6

I will also write a quick post giving the reasons why I opted for that approach. 
Feel free to give any other constructive feedback. 😄

**Comment by android10** [comment]
I also wrote an article with the arguments behind the design decisions:
http://fernandocejas.com/2016/12/24/clean-architecture-dynamic-parameters-in-use-cases/

As usual, any feedback is very welcome. Merry Christmas! 😄 🎄

**Comment by Dharmendra10** [comment]
I know we have few alternative for the dynamic argument for the UseCase but I want to share my approach. This is similar approach like @roscrazy just little changes.

Here is UseCase

```
public class GetUserUseCase extends UseCase {

    private final UserRepository userRepository;
    private String userId;

    @Inject
    public GetUserUseCase(UserRepository userRepository, ThreadExecutor threadExecutor, PostExecutionThread postExecutionThread) {
        super(threadExecutor, postExecutionThread);
        this.userRepository = userRepository;
    }

    public void setUserId(String userId) {
        this.userId = userId;
    }

    @Override
    public Observable buildUseCaseObservable() {
        return this.userRepository.getUser(userId);
    }
}
```
Here is the Factory class for the User
```
@Singleton
public class UserUseCaseFactory {
    @Inject
    public CartUseCaseFactory(UserRepository userRepository, ThreadExecutor threadExecutor, PostExecutionThread postExecutionThread) {
        this.userRepository = userRepository;
        this.threadExecutor = threadExecutor;
        this.postExecutionThread = postExecutionThread;
    }

    public UseCase provideGetUserDetailUseCase(String userId) {
        GetUserUseCase mGetUserUseCase = new GetUserUseCase(userRepository, threadExecutor, postExecutionThread);
        mGetUserUseCase.setUserId(userId);
        return mGetUserUseCase;
    }
}
```

Inside presenter the factory will provide the UseCase

```
public void loadUser(String userId) {
        getUserUseCase = mUserUseCaseFactory.provideGetUserDetailUseCase(userId);
        getUserUseCase.execute(new GetUserSubscriber());
}
```

**Comment by android10** [comment]
@Dharmendra10 We know there are no silver bullets and sharing experiences is awesome. Thanks

**Comment by AnupAmmanavar** [comment]
Kotlin here.

create an interface (here : `Parameters`) for holding the params.
create classes (or data classes) which extend the interface to hold the data, as shown below.

```
interface Parameters

class LoginParams(val userName: String, val password: String) : Parameters

class GetUserDetailsParams(val id: Long): Parameters

```
Change your base `Usecase`(here : `ObservableUseCase`) to take in the `Parameter` interface to make it generic.

```
abstract class ObservableUseCase<T, P : Parameters> constructor(
    private val postExecutionThread: PostExecutionThread
) {

     protected abstract fun buildUseCaseObservable(p: P): Observable<T>

    open fun invoke(singleObserver: DisposableObserver<T>, p: P) {
        val single = this.buildUseCaseObservable(p)
            .subscribeOn(Schedulers.io())
            .observeOn(postExecutionThread.scheduler)
    }
}
```


According the changes in the  `usecase` look like this

```
open class LoginUseCase @Inject constructor(
    postExecutionThread: PostExecutionThread,
    private val userRepository: UserRepository
) : ObservableUseCase<User, LoginParams>(postExecutionThread) {

    override fun buildUseCaseObservable(arguments: LoginParams): Observable<User> {
        return userRepository.login(arguments.userName, arguments.password)
    }
}
```

If your usecase doesnot take in any arguments then you can use this.

```
// This is useful when your use case needs no parameters
class Empty : Parameters

```

**Comment by jamolkhon** [comment]
What's the point of having a UseCase interface/base class? What's wrong with a plain class like this:
```
class LoginUseCase @Inject constructor(private val loginService: LoginService) {
  fun login(username: String, password: String): Single<LoginResult> {
    ...
  }
}
```

**Comment by MaxNF** [comment]
> What's the point of having a UseCase interface/base class? What's wrong with a plain class like this:
> 
> ```
> class LoginUseCase @Inject constructor(private val loginService: LoginService) {
>   fun login(username: String, password: String): Single<LoginResult> {
>     ...
>   }
> }
> ```

I use a plain approach like the one you wrote here. It is flexible and works very well for me. But this is mainly because my use cases do not use shared code, which should be placed in a super class. But I guess in really big projects with several developers a more strict approach (with a base class or an interface) can be useful, but I have not encountered this necessity yet.
